### PR TITLE
[GEOS-9805] Support transforming CAS response custom attributes into roles

### PR DESCRIPTION
**Pull request by aaime**
See : https://osgeo-org.atlassian.net/browse/GEOS-9805

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [ ] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [ ] PR for bug fixes and small new features are presented as a single commit
- [ ] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [ ] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [ ] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [ ] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [ ] Commits changing the REST API, or any configuration object, should check if the REST API docs (Swagger YAML files and classic documentation) need to be updated.

### CODE REVIEW

**Comment by aaime** [comment]
The doc failure happens only on Travis, looks like an outdated python/sphinx combination:

```
     [echo] Running sphinx-build -D release=2.19-SNAPSHOT -W -b html -d "/home/travis/build/geoserver/geoserver/doc/en/target/docguide/doctrees" . "/home/travis/build/geoserver/geoserver/doc/en/target/docguide/html"
     [exec] /usr/local/lib/python2.7/dist-packages/sphinx/util/nodes.py:94: FutureWarning: 
     [exec]    The iterable returned by Node.traverse()
     [exec]    will become an iterator instead of a list in Docutils > 0.16.
     [exec]   for classifier in reversed(node.parent.traverse(nodes.classifier)):
     [exec] /usr/local/lib/python2.7/dist-packages/sphinx/writers/html.py:462: FutureWarning: 
     [exec]    The iterable returned by Node.traverse()
     [exec]    will become an iterator instead of a list in Docutils > 0.16.
     [exec]   target_node = image_nodes and image_nodes[0] or node.parent
```

Locally it builds fine:

```
init:
     [echo] Base directory is /home/aaime/devel/git-gs/doc/en
     [echo] Build directory is /home/aaime/devel/git-gs/doc/en/target and project version is 2.19-SNAPSHOT

docguide:

sphinx:
     [echo] Running sphinx-build -D release=2.19-SNAPSHOT -W -b html -d "/home/aaime/devel/git-gs/doc/en/target/docguide/doctrees" . "/home/aaime/devel/git-gs/doc/en/target/docguide/html"

user:

sphinx:
     [echo] Running sphinx-build -D release=2.19-SNAPSHOT -W -b html -d "/home/aaime/devel/git-gs/doc/en/target/user/doctrees" . "/home/aaime/devel/git-gs/doc/en/target/user/html"

developer:

sphinx:
     [echo] Running sphinx-build -D release=2.19-SNAPSHOT -W -b html -d "/home/aaime/devel/git-gs/doc/en/target/developer/doctrees" . "/home/aaime/devel/git-gs/doc/en/target/developer/html"

````

File: src/extension/security/cas/src/main/java/org/geoserver/security/cas/GeoServerCasAuthenticationFilter.java [hunk 1]
```diff
@@ -258,4 +270,63 @@ public void logout(
             Authentication authentication) {
         request.setAttribute(GeoServerLogoutFilter.LOGOUT_REDIRECT_ATTR, casLogoutURL);
     }
+
+    @Override
+    protected Collection<GeoServerRole> getRoles(HttpServletRequest request, String principal)
```
Comment (afabiani) [inline src/extension/security/cas/src/main/java/org/geoserver/security/cas/GeoServerCasAuthenticationFilter.java:275]: Not really and issue, but for completeness I guess that would be good to complete the principal roles list by using the RoleCalculator

```
RoleCalculator calc =
                        new RoleCalculator(getSecurityManager().getActiveRoleService());
```

In order to be sure we add all the roles associated to the current user accordingly to the active role service too.

**Review by afabiani — APPROVED** [review]
Except for one minor observation, the PR looks good to me.

File: src/extension/security/cas/src/main/java/org/geoserver/security/cas/GeoServerCasAuthenticationFilter.java (hunk 1 above)
Comment (aaime) [reply src/extension/security/cas/src/main/java/org/geoserver/security/cas/GeoServerCasAuthenticationFilter.java:275]: Interesting, I had no idea this class existed. The base class does not do anything with the RoleCalculator, but maybe it should not be done when picking roles from a specific role service or a specific user group service. 
Reading from the headers does nothing either, but I see the J2EE auth (another subclass of the pre authenticated filter) using it.... adding it in the "roles from custom attributes" path only, don't want to touch the base classes.

**Comment by aaime** [comment]
Failure in JDK11 is unrelated and due to a time based test, will handle it separately.

**Pull request merged**